### PR TITLE
fix : Username display after login (#166)

### LINKED ISSUE #166: Bug: Display email issue
**Describe the bug**

When a user logs in using username it display message "Welcome ${username}" and when the same user logs in using email it displays "Welcome ${email}" 

**Expected behavior**

Either the user logs in using email or username ,It should show user's username.

**Screenshots**

Peek 2021-02-01 02-23
**Desktop (please complete the following information):**
 - OS: any
 - Browser  any
 - Version  any

**Smartphone (please complete the following information):**
 - Device: any
 - OS:  any
 - Browser  any
 - Version  any

### PR DESCRIPTION
**Pull request by Aryamanz29**
### Description
 
Now username can be seen over navbar.

Fixes #166

### Type of Change:

- Code and UI

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?

issueUsername

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] I have written Kotlin Docs whenever is applicable

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Review by devkapilbansal — COMMENTED** [review]
Hi @Aryamanz29 Although the changes made by you may work but I think we can make changes to navigation bar only to display user's name

**Comment by Aryamanz29** [comment]
> Hi @Aryamanz29 Although the changes made by you may work but I think we can make changes to navigation bar only to display user's name

Hello @devkapilbansal  ,  I didn't find another way other than fetching through `user/personal_details` endpoint because earlier in  `navigation.js, useContext(AuthContext)` returning user of type string which depends upon whether the user uses email/username while login which creates problem there. Also, this problem is related to the backend only that's why fetching details from `BIT API` is a better solution in my opinion. I'm a beginner with react & jsx . Any suggestions regarding this will be highly appreciated.

**Comment by decon-harsh** [comment]
> Hi @Aryamanz29 Although the changes made by you may work but I think we can make changes to navigation bar only to display user's name

Hey @devkapilbansal are you suggesting BIT back-end /login should return the username on success as well?

**Comment by Rahulm2310** [comment]
@devkapilbansal @decon-harsh @Aryamanz29  Won't it better to store the whole user object in the auth context rather than storing a username string. It might help us avoid API calls for user personal details on other pages as well.

**Comment by decon-harsh** [comment]
> @devkapilbansal @decon-harsh @Aryamanz29  Won't it better to store the whole user object in the auth context rather than storing a username string. It might help us avoid API calls at other pages as well. 

On success /login should return UserBody (except password ofc)? Right @Rahulm2310 ?

File: src/login/Login.jsx [hunk 1]
```diff
@@ -8,7 +8,7 @@ import { SERVICE_UNAVAILABLE_ERROR } from "../messages";
 export default function Login() {
     const [errorMessage, setErrorMessage] = useState(null);
     const {isAuth,login} = useContext(AuthContext);
-    const [user, setUser] = useState(null);
+    const [, setUser] = useState(null);
```
Comment (codesankalp) [inline src/login/Login.jsx]: don't remove this, you can use this to store the username also.

**Comment by codesankalp** [comment]
I saw the code and it is storing the user from the form data instead of API response while login so I think this needs to be addressed in the backend because the login endpoint must return user info like username, email with access_token, and refresh_token. So that it will be easy to get user info without calling another API.
I think it needs to be addressed in the backend as well. What do you think @Rahulm2310 @decon-harsh @devkapilbansal @Aryamanz29

**Comment by decon-harsh** [comment]
> I saw the code and it is storing the user from the form data instead of API response while login so I think this needs to be addressed in the backend because the login endpoint must return user info like username, email with access_token, and refresh_token. So that it will be easy to get user info without calling another API.
> I think it needs to be addressed in the backend as well. What do you think @Rahulm2310 @decon-harsh @devkapilbansal @Aryamanz29 

Why shouldn't it return full user api model (except password)?

**Comment by codesankalp** [comment]
> > I saw the code and it is storing the user from the form data instead of API response while login so I think this needs to be addressed in the backend because the login endpoint must return user info like username, email with access_token, and refresh_token. So that it will be easy to get user info without calling another API.
> > I think it needs to be addressed in the backend as well. What do you think @Rahulm2310 @decon-harsh @devkapilbansal @Aryamanz29
> 
> Why shouldn't it return full user api model (except password)?

Because there is another end-point for this.

**Comment by decon-harsh** [comment]
@codesankalp yes you are right. /login should return username and email along with success message. 


If admins and other maintainers agree we can open this issue on BIT back-end and till then keep this on hold?

**Comment by mtreacy002** [comment]
> I saw the code and it is storing the user from the form data instead of API response while login so I think this needs to be addressed in the backend because the login endpoint must return user info like username, email with access_token, and refresh_token. So that it will be easy to get user info without calling another API.

@codesankalp and @decon-harsh , if we are to go in this direction then the issue should be raised on MS backend since BIT just returning whatever being returned by the MS backend login response model 😉. IMO, I don't think  it's a good idea to modify the login response model only at BIT backend bcoz  managing users is within the jurisdiction of MS backend. BIT can only modify users api models related to user's additional info and personal background which are specific to BIT. What do you think?

**Comment by mtreacy002** [comment]
> @devkapilbansal @decon-harsh @Aryamanz29 Won't it better to store the whole user object in the auth context rather than storing a username string. It might help us avoid API calls for user personal details on other pages as well.

@Rahulm2310 , for this, we need to make a second api call to GET /user/personal_details endpoint after the first api call to POST /login is successful. Then we can save full user's data in cookies instead of just username/email taken from the login form field. However, keeping this data in cookies might add another security concern on top of what I've already raised on [BIT project ideas](https://github.com/anitab-org/bridge-in-tech-backend/wiki/Project-ideas#backend--frontend-full-features) for keeping access_token as cookies. This is also why atm we only grab user's personal details when we need it which is when we view `personal details` page. None of the other pages we have atm need user's data in details.
Screen Shot 2021-04-23 at 10 23 29 am
I'd like to hear your opinion on this 😉

**Comment by devkapilbansal** [comment]
> @codesankalp yes you are right. /login should return username and email along with success message.
> 
> If admins and other maintainers agree we can open this issue on BIT back-end and till then keep this on hold?

@mtreacy002 I agree with what @decon-harsh suggested here

**Comment by mtreacy002** [comment]
> > @codesankalp yes you are right. /login should return username and email along with success message.
> > If admins and other maintainers agree we can open this issue on BIT back-end and till then keep this on hold?
> 
> @mtreacy002 I agree with what @decon-harsh suggested here

@devkapilbansal , so you're suggesting that we should open an issue on MS backend to change the return from POST /login endpoint, right? (following the reason I stated [above](https://github.com/anitab-org/bridge-in-tech-web/pull/235#issuecomment-825280821))

**Comment by devkapilbansal** [comment]
> > > @codesankalp yes you are right. /login should return username and email along with success message.
> > > If admins and other maintainers agree we can open this issue on BIT back-end and till then keep this on hold?
> > 
> > 
> > @mtreacy002 I agree with what @decon-harsh suggested here
> 
> @devkapilbansal , so you're suggesting that we should open an issue on MS backend to change the return from POST /login endpoint, right? (following the reason I stated [above](https://github.com/anitab-org/bridge-in-tech-web/pull/235#issuecomment-825280821))

Yes @mtreacy002 Although I don't know whether it would be good to do or not

**Comment by mtreacy002** [comment]
> > > > @codesankalp yes you are right. /login should return username and email along with success message.
> > > > If admins and other maintainers agree we can open this issue on BIT back-end and till then keep this on hold?
> > > 
> > > 
> > > @mtreacy002 I agree with what @decon-harsh suggested here
> > 
> > 
> > @devkapilbansal , so you're suggesting that we should open an issue on MS backend to change the return from POST /login endpoint, right? (following the reason I stated [above](https://github.com/anitab-org/bridge-in-tech-web/pull/235#issuecomment-825280821))
> 
> Yes @mtreacy002 Although I don't know whether it would be good to do or not

@devkapilbansal, I think we should ask @isabelcosta 's opinion on this 😉 . But for the purpose of this PR, we need to limit the scope to whether or not username is shown on the navbar when user logged in either with email or username. I believe @Aryamanz29 has completed the required task in this sense. We can move discussion on the other issue (to add username/email in the Login response model) on MS backend's Zulip channel and open an issue once @isabelcosta gives her approval. What do you think? If you agree, can you please approve the PR? Thanks

File: src/login/Login.jsx (hunk 1 above)
Comment (mtreacy002) [reply src/login/Login.jsx]: @Aryamanz29, can you please submit the new change as requested by @codesankalp here?

Comment (Aryamanz29) [reply src/login/Login.jsx]: @mtreacy002 Done 😄

**Review by codesankalp — APPROVED** [review]
As the username display is fixed after login, I am approving this PR as mentioned by @mtreacy002.

**Review by Rahulm2310 — APPROVED** [review]
Looks good :100:.Thanks @Aryamanz29 for your contribution :tada: 
@mtreacy002 don't we need to test this PR?

**Review by mtreacy002 — APPROVED** [review]
Well done for the fix, @Aryamanz29  👍 
I've tested this locally and here's the test summary:
1. Code review: Done
2. Tested all possible outcomes
    Expected behaviour: Either user logged in with username or email, only username should be shown on the main navbar (Welcome, {username}!)
    Actual behaviour: as expected (see gifs below)
* when user logged in with username
ezgif com-gif-maker (54)

* when user logged in with their email 
ezgif com-gif-maker (55)

3. Environments:
    OS: MacOS Catalina v 10.15.7
    Browser: Google Chrome

Thank you again, @Aryamanz29 for your contribution 🎉 🎉 🎉

**Pull request merged**